From e3bf8850185cfa16132c5574a40e6e92973973d0 Mon Sep 17 00:00:00 2001
From: Xuefu Zhang <xzhang@Cloudera.com>
Date: Wed, 20 May 2015 11:05:07 -0700
Subject: [PATCH 0305/1164] CDH-31906: HIVE-8529: HiveSessionImpl#fetchResults
 should not try to fetch operation log when
 hive.server2.logging.operation.enabled is false
 (Yongzhi via Xuefu)

---
 .../java/org/apache/hive/jdbc/TestJdbcDriver2.java |   20 ++++++++++++++++++++
 .../service/cli/operation/OperationManager.java    |   13 ++++++++++---
 .../hive/service/cli/session/HiveSessionImpl.java  |    2 +-
 3 files changed, 31 insertions(+), 4 deletions(-)

diff --git a/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java b/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java
index 2c85877..c32479a 100644
--- a/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java
+++ b/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java
@@ -2284,6 +2284,26 @@ public void run() {
     verifyFetchedLog(incrementalLogs, expectedLogs);
   }
 
+  /**
+   * Test getting query log when HS2 disable logging.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testGetQueryLogOnDisabledLog() throws Exception {
+    Statement setStmt = con.createStatement();
+    setStmt.execute("set hive.server2.logging.operation.enabled = false");
+    String sql = "select count(*) from " + tableName;
+    HiveStatement stmt = (HiveStatement)con.createStatement();
+    assertNotNull("Statement is null", stmt);
+    stmt.executeQuery(sql);
+    List<String> logs = stmt.getQueryLog(false, 10);
+    stmt.close();
+    assertTrue(logs.size() == 0);
+    setStmt.execute("set hive.server2.logging.operation.enabled = true");
+    setStmt.close();
+  }
+
   private void verifyFetchedLog(List<String> logs, String[] expectedLogs) {
     StringBuilder stringBuilder = new StringBuilder();
 
diff --git a/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java b/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java
index 6cae8a8..52c2299 100644
--- a/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java
+++ b/service/src/java/org/apache/hive/service/cli/operation/OperationManager.java
@@ -28,6 +28,7 @@
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.conf.HiveConf.ConfVars;
 import org.apache.hadoop.hive.metastore.api.FieldSchema;
 import org.apache.hadoop.hive.metastore.api.Schema;
 import org.apache.hadoop.hive.ql.session.OperationLog;
@@ -251,14 +252,22 @@ public RowSet getOperationNextRowSet(OperationHandle opHandle,
   }
 
   public RowSet getOperationLogRowSet(OperationHandle opHandle,
-      FetchOrientation orientation, long maxRows)
+      FetchOrientation orientation, long maxRows, HiveConf hConf)
           throws HiveSQLException {
+    TableSchema tableSchema = new TableSchema(getLogSchema());
+    RowSet rowSet = RowSetFactory.create(tableSchema, getOperation(opHandle).getProtocolVersion());
+
+    if (hConf.getBoolVar(ConfVars.HIVE_SERVER2_LOGGING_OPERATION_ENABLED) == false) {
+      LOG.warn("Try to get operation log when hive.server2.logging.operation.enabled is false, no log will be returned. ");
+      return rowSet;
+    }
     // get the OperationLog object from the operation
     OperationLog operationLog = getOperation(opHandle).getOperationLog();
     if (operationLog == null) {
       throw new HiveSQLException("Couldn't find log associated with operation handle: " + opHandle);
     }
 
+
     // read logs
     List<String> logs;
     try {
@@ -269,8 +278,6 @@ public RowSet getOperationLogRowSet(OperationHandle opHandle,
 
 
     // convert logs to RowSet
-    TableSchema tableSchema = new TableSchema(getLogSchema());
-    RowSet rowSet = RowSetFactory.create(tableSchema, getOperation(opHandle).getProtocolVersion());
     for (String log : logs) {
       rowSet.addRow(new String[] {log});
     }
diff --git a/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java b/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
index f29a8dd..0280677 100644
--- a/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
+++ b/service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java
@@ -731,7 +731,7 @@ public RowSet fetchResults(OperationHandle opHandle, FetchOrientation orientatio
       if (fetchType == FetchType.QUERY_OUTPUT) {
         return operationManager.getOperationNextRowSet(opHandle, orientation, maxRows);
       }
-      return operationManager.getOperationLogRowSet(opHandle, orientation, maxRows);
+      return operationManager.getOperationLogRowSet(opHandle, orientation, maxRows, hiveConf);
     } finally {
       release(true);
     }
-- 
1.7.9.5

